### PR TITLE
feat: add telemetry for CairoEngine events + PGWire and HTTP JSON apis

### PR DESCRIPTION
**Pull request by mpsq**
Ideas:
- Remove `telemetry` / `telemetry_config` from `SHOW TABLES;`
- Add tests for PGWire

### CODE REVIEW

File: core/src/main/java/io/questdb/PropServerConfiguration.java [hunk 1]
```diff
@@ -224,7 +226,7 @@
     private boolean pgDaemonPool;
 
     public PropServerConfiguration(String root, Properties properties) throws ServerConfigurationException, JsonException {
-        this.sharedWorkerCount = getInt(properties, "shared.worker.count", 2);
+        this.sharedWorkerCount = getInt(properties, "shared.worker.count", 3);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/PropServerConfiguration.java:229]: this looks like an unnecessary change

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 2]
```diff
@@ -63,6 +66,10 @@ public CairoEngine(CairoConfiguration configuration, @Nullable MessageBus messag
         this.writerPool = new WriterPool(configuration, messageBus);
         this.readerPool = new ReaderPool(configuration);
         this.writerMaintenanceJob = new WriterMaintenanceJob(configuration);
+        this.telemetryQueue = new RingQueue<>(TelemetryRow::new, configuration.getTelemetryQueueCapacity());
+        this.telemetryPubSeq = new SPSequence(configuration.getTelemetryQueueCapacity());
+        this.telemetrySubSeq = new MCSequence(configuration.getTelemetryQueueCapacity());
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/CairoEngine.java:71]: this is perhaps incorrect. This queue should be Multiple publisher single consumer

File: core/src/main/java/io/questdb/griffin/CompiledQuery.java [hunk 3]
```diff
@@ -29,27 +29,27 @@
 import io.questdb.cutlass.text.TextLoader;
 
 public interface CompiledQuery {
-    int SELECT = 0;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/CompiledQuery.java:32]: why change?

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 4]
```diff
@@ -310,6 +318,108 @@ private void rename0(Path path, CharSequence tableName, Path otherPath, CharSequ
...
+            return Long.toHexString(hi | (val & (hi - 1))).substring(1);
+        }
+
+        private void addEvent(short event) {
+            final TelemetryRow row = new TelemetryRow();
+
+            row.ts = microsecondClock.getTicks();
+            row.id = id;
+            row.event = event;
+            toTelemetryTable(row);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/CairoEngine.java:395]: storing own events can be done via publishing to the queue, it will make handling of events more consistent and avoid allocating object instances to pass method args

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 5]
```diff
@@ -310,6 +318,108 @@ private void rename0(Path path, CharSequence tableName, Path otherPath, CharSequ
...
+        }
+
+        private void toTelemetryTable(TelemetryRow telemetryRow) {
+            final TableWriter.Row row = writer.newRow();
+
+            row.putDate(0, telemetryRow.ts);
+            row.putStr(1, id);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/CairoEngine.java:402]: storing ID on every telemetry row as string is not efficient. The ID can be stored once in separate table and cross-joined to the time series when getting data out

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 6]
```diff
@@ -310,6 +318,108 @@ private void rename0(Path path, CharSequence tableName, Path otherPath, CharSequ
...
+        private final QueueConsumer<TelemetryRow> myConsumer = this::toTelemetryTable;
+        private final CharSequence id;
+        private final TableWriter writer;
+
+        public TelemetryWriterJob(CairoConfiguration configuration) {
+            final CharSequence root = configuration.getRoot();
+
+            try(Path path = new Path()) {
+                if (getStatus(AllowAllCairoSecurityContext.INSTANCE, path, telemetryTableName) == TableUtils.TABLE_DOES_NOT_EXIST) {
+                    final TelemetryTableModel telemetry = new TelemetryTableModel(telemetryTableName);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/CairoEngine.java:350]: i should have mentioned this earlier, sorry, but table can be created via SqlCompiler and SQL, way simpler. We can pass the call site via SqlExecution context, which is everywhere CairoEngine is.

At this juncture I would suggest we create another table where we store guid
the guid should be of Long256 type

File: core/src/main/java/io/questdb/griffin/CompiledQuery.java (hunk 3 above)
Comment (mpsq) [reply core/src/main/java/io/questdb/griffin/CompiledQuery.java:32]: Because ultimately the event that I store is a short. I looked at this and all we use is 1 to 14, I thought we might save the extra space and use a short too since it is very unlikely we will ever need bigger than that. I can revert though, let me know.

Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/CompiledQuery.java:32]: It didn’t click with me at the time that these are the values you store. It is ok to change type!

File: core/src/main/java/io/questdb/cairo/CairoEngine.java (hunk 4 above)
Comment (mpsq) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java:395]: done

File: core/src/main/java/io/questdb/PropServerConfiguration.java (hunk 1 above)
Comment (mpsq) [reply core/src/main/java/io/questdb/PropServerConfiguration.java:229]: reverted

File: core/src/main/java/io/questdb/cairo/CairoEngine.java (hunk 5 above)
Comment (mpsq) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java:402]: done

File: core/src/main/java/io/questdb/cairo/CairoEngine.java (hunk 6 above)
Comment (mpsq) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java]: done

**Comment by mpsq** [comment]
All the comments were addressed :) @bluestreak01

File: core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessor.java [hunk 7]
```diff
@@ -310,6 +311,7 @@ private static void syntaxError(
     private void compileQuery(JsonQueryProcessorState state) throws SqlException, PeerDisconnectedException, PeerIsSlowToReadException {
         final long nanos = nanosecondClock.getTicks();
         final CompiledQuery cc = compiler.compile(state.getQuery(), sqlExecutionContext);
+        compiler.getEngine().storeTelemetry(cc.getType());
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessor.java:314]: lets pass the call site of storeTelemetry() via sqlExecuitionContext rather than CairoEngine

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 8]
```diff
@@ -318,6 +334,101 @@ private void rename0(Path path, CharSequence tableName, Path otherPath, CharSequ
...
+                    final Long256Impl id = new Long256Impl();
+                    id.setLong0(nanosecondClock.getTicks());
+                    id.setLong1(microsecondClock.getTicks());
+                    id.toSink(idSink);
+                    compiler.compile("CREATE TABLE " + telemetryConfigTableName + " (id long256, enabled boolean);", sqlExecutionContext);
+                } else {
+                    try (TableReader reader = new TableReader(configuration, telemetryConfigTableName)) {
+                        reader.getCursor().getRecord().getLong256(0, idSink);
+                    }
+                    compiler.compile("TRUNCATE TABLE " + telemetryConfigTableName + ";", sqlExecutionContext);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/CairoEngine.java:384]: Semicolon is optional.

As discussed in-person we probably should keep config table. In fact two:

1. table for guid - created once if it doesn't exist
2. table for ups and downs. I would probably put up/down events to main telemetry table?

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 9]
```diff
@@ -318,6 +334,101 @@ private void rename0(Path path, CharSequence tableName, Path otherPath, CharSequ
...
+        public TelemetryWriterJob(CairoConfiguration configuration) throws SqlException {
+            sqlExecutionContext.with(AllowAllCairoSecurityContext.INSTANCE, null, null);
+
+            try(Path path = new Path()) {
+                final NanosecondClock nanosecondClock = configuration.getNanosecondClock();
+                final MicrosecondClock microsecondClock = configuration.getMicrosecondClock();
+
+                if (getStatus(AllowAllCairoSecurityContext.INSTANCE, path, telemetryConfigTableName) == TableUtils.TABLE_DOES_NOT_EXIST) {
+                    final Long256Impl id = new Long256Impl();
+                    id.setLong0(nanosecondClock.getTicks());
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/CairoEngine.java:376]: there is a way to set long256 on table writer once it is opened. This would make creating Long256Impl unnecessary.

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 10]
```diff
@@ -318,6 +334,101 @@ private void rename0(Path path, CharSequence tableName, Path otherPath, CharSequ
...
+                    compiler.compile("TRUNCATE TABLE " + telemetryConfigTableName + ";", sqlExecutionContext);
+                }
+
+                if (getStatus(AllowAllCairoSecurityContext.INSTANCE, path, telemetryTableName) == TableUtils.TABLE_DOES_NOT_EXIST) {
+                    compiler.compile("CREATE TABLE " + telemetryTableName + " (ts timestamp, event short);", sqlExecutionContext);
+                }
+            }
+
+            this.writer = new TableWriter(configuration, telemetryTableName);
+            this.writerConfig = new TableWriter(configuration, telemetryConfigTableName);
```
[the remark this answers is not in the document]
Comment (mpsq) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java:393]: The reason why I kept the writer is to prevent user from tweaking the table

File: core/src/main/java/io/questdb/cairo/CairoEngine.java (hunk 8 above)
Comment (mpsq) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java:384]: I don't understand this, this is already the case:

1. We have a table `telemetry_config` that stores `guid` and `enabled`
2. We do exactly that

File: core/src/main/java/io/questdb/cairo/CairoEngine.java [hunk 11]
```diff
@@ -318,6 +334,101 @@ private void rename0(Path path, CharSequence tableName, Path otherPath, CharSequ
...
+            row.append();
+        }
+
+        @Override
+        public boolean runSerially() {
+            if (!telemetryEnabled) {
+                return true;
+            }
+
+            telemetrySubSeq.consumeAll(telemetryQueue, myConsumer);
```
[the remark this answers is not in the document]
Comment (mpsq) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java:417]: What I understand is:
- if telemetry is disabled return false
- when enabled, return the value of `telemetrySubSeq.consumeAll(...)`

It that correct?

File: core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessor.java (hunk 7 above)
Comment (mpsq) [reply core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessor.java:317]: If I pass the executionContext this will be potentially have the a security context that do not allow writes (when read only), by using my own context I can use:

```
sqlExecutionContext.with(AllowAllCairoSecurityContext.INSTANCE, null, null);
```

File: core/src/main/java/io/questdb/cairo/CairoEngine.java (hunk 9 above)
Comment (mpsq) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java:376]: Done, it makes the code somehow convoluted, let me know what you think

File: core/src/main/java/io/questdb/cairo/CairoEngine.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/CairoEngine.java:417]: Return `false` in both cases. `true` means that scheduler will rush to call you again. If you consumed queue fully there is probably no rush to consume it again. Generally this should be the case anyway.

File: core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessor.java (hunk 7 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessor.java:317]: you can have the same security context as assigned to the execution context :)
